numpy/core/arrayprint.py

```diff
@@ -641,53 +647,58 @@ def fillFormat(self, data):
-            self.max_str_len = 8 + self.precision
-            if self.large_exponent:
-                self.max_str_len += 1
-            if self.sign:
-                format = '%+'
-            else:
-                format = '%'
-            format = format + '%d.%de' % (self.max_str_len, self.precision)
+
+            signpos = self.sign != '-' or _nc.any(non_zero < 0)
+            # for back-compatibility with np 1.13, add extra space if padded
+            signpos = signpos if self.sign != ' ' else 2
+            max_str_len = signpos + 6 + self.precision + self.large_exponent
+
+            signchar = '' if self.sign == '-' else self.sign
+            format = '%' + signchar + '%d.%de' % (max_str_len, self.precision)
         else:
-            format = '%%.%df' % (self.precision,)
-            if len(non_zero):
-                precision = max([_digits(x, self.precision, format)
-                                 for x in non_zero])
+            if len(non_zero) and self.precision > 0:
+                precision = self.precision
+                trim_zero = lambda s: precision - (len(s) - len(s.rstrip('0')))
+                fmt = '%%.%df' % (precision,)
+                precision = max(trim_zero(fmt % x) for x in abs_non_zero)
             else:
                 precision = 0
-            precision = min(self.precision, precision)
-            self.max_str_len = len(str(int(max_val))) + precision + 2
+
+            int_len = len(str(int(max_val)))
+            signpos = self.sign != '-' or (len(str(int(min_val_sgn))) > int_len)
+            max_str_len = signpos + int_len + 1 + precision
+
             if _nc.any(special):
-                self.max_str_len = max(self.max_str_len,
-                                       len(_nan_str),
-                                       len(_inf_str)+1)
-            if self.sign:
-                format = '%#+'
-            else:
-                format = '%#'
-            format = format + '%d.%df' % (self.max_str_len, precision)
+                neginf = self.sign != '-' or _nc.any(data[hasinf] < 0)
+                nanlen = len(_format_options['nanstr'])
+                inflen = len(_format_options['infstr']) + neginf
+                max_str_len = max(max_str_len, nanlen, inflen)
+
+            signchar = '' if self.sign == '-' else self.sign
+            format = '%#' + signchar + '%d.%df' % (max_str_len, precision)
 
-        self.special_fmt = '%%%ds' % (self.max_str_len,)
+        self.special_fmt = '%%%ds' % (max_str_len,)
         self.format = format
 
     def __call__(self, x, strip_zeros=True):
         from . import numeric as _nc
 
         with _nc.errstate(invalid='ignore'):
             if isnan(x):
-                if self.sign:
-                    return self.special_fmt % ('+' + _nan_str,)
+                nan_str = _format_options['nanstr']
+                if self.sign == '+':
+                    return self.special_fmt % ('+' + nan_str,)
```

Comment:
Not your fault, but yikes, since when is `nan` known to be positive? Can we omit this?

Reply:
Hmm removing this actuallt breaks tests, because complex types require a sign in the imag part: `0.+nanj`

Reply:
I guess to avoid that one would need yet another special-case for "really, I always want a sign", which is used for complex. But that seems overly cumbersome for a case where there is no confusion, so let's just let it be.